### PR TITLE
Add cloudbuild.yaml and update Makefile

### LINKED ISSUE #47: Investigate solution to CRD being too large to use `kubectl apply`
Currently the `make install` and `make deploy` commands won't work if the CRD / controller already exist on the cluster, since instead of doing `kubectl apply` they are both doing `kubectl create` now since the CRD is too large for `kubectl apply` as it runs into this [issue](https://github.com/kubernetes-sigs/kubebuilder/issues/1140). 

I'm guessing this will be an unexpected user experience for users so we need to figure how will someone update their existing deployment without interruptions if only kubectl create works (implying they must delete then recreate). Alternatively, we need to figure out how to reduce the size of the CRD.

### PR DESCRIPTION
**Pull request by danielvegamyhre**
Part of #22
Fixed #47

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -73,6 +98,30 @@ build: manifests generate fmt vet ## Build manager binary.
 run: manifests generate fmt vet ## Run a controller from your host.
 	go run ./main.go
 
+# Build the container image
```
Comment (ahg-g) [inline Makefile:101]: replace the above `go` calls with $(GO_CMD)

File: Makefile [hunk 2]
```diff
@@ -3,15 +3,40 @@ IMG ?= controller:latest
...
-
 # Get the currently used golang install path (in GOPATH/bin, unless GOBIN is set)
 ifeq (,$(shell go env GOBIN))
 GOBIN=$(shell go env GOPATH)/bin
 else
 GOBIN=$(shell go env GOBIN)
 endif
 
+GO_CMD ?= go
+GO_FMT ?= gofmt
```
Comment (ahg-g) [inline Makefile:14]: not used, perhaps meant to replace `go fmt`?

File: Makefile [hunk 3]
```diff
@@ -3,15 +3,40 @@ IMG ?= controller:latest
...
+GO_TEST_FLAGS ?= -race
```
Comment (ahg-g) [inline Makefile:15]: not used

File: Makefile [hunk 4]
```diff
@@ -3,15 +3,40 @@ IMG ?= controller:latest
...
+GIT_TAG ?= $(shell git describe --tags --dirty --always)
+# Image URL to use all building/pushing image targets
+PLATFORMS ?= linux/amd64,linux/arm64
+DOCKER_BUILDX_CMD ?= docker buildx
+IMAGE_BUILD_CMD ?= $(DOCKER_BUILDX_CMD) build
+IMAGE_BUILD_EXTRA_OPTS ?=
+IMAGE_REGISTRY ?= gcr.io/k8s-staging-jobset
+IMAGE_NAME := jobset
+IMAGE_REPO ?= $(IMAGE_REGISTRY)/$(IMAGE_NAME)
+IMAGE_TAG ?= $(IMAGE_REPO):$(GIT_TAG)
```
Comment (ahg-g) [inline Makefile:26]: there are multiple reference to `${IMG}`, I think we should replace them with ${IMAGE_TAG}`

File: Makefile [hunk 5]
```diff
@@ -73,6 +98,30 @@ build: manifests generate fmt vet ## Build manager binary.
...
+		$(PUSH) \
+		$(IMAGE_BUILD_EXTRA_OPTS) ./
+
+.PHONY: image-push
+image-push: PUSH=--push
+image-push: image-build
+
 # If you wish built the manager image targeting other platforms you can use the --platform flag.
 # (i.e. docker build --platform linux/arm64 ). However, you must enable docker buildKit for it.
 # More info: https://docs.docker.com/develop/develop-images/build_enhancements/
```
Comment (ahg-g) [inline Makefile:127]: do we need docker-build, docker-buildx and docker-push now that we have the new rules above?

File: cloudbuild.yaml [hunk 6]
```diff
@@ -0,0 +1,27 @@
+
+ 
+
```
Comment (ahg-g) [inline cloudbuild.yaml:3]: remove empty lines

File: Makefile (hunk 5 above)
Comment (ahg-g) [reply Makefile:127]: yes, lets update the readme at the same time, like below; the image-local-push rule should be testable before you submit

```
git clone https://github.com/kubernetes-sigs/jobset.git
cd jobset
IMAGE_REGISTRY=registry.example.com/my-user make image-local-push deploy
```

Comment (danielvegamyhre) [reply Makefile]: Tested new build+deploy manually using `IMAGE_REGISTRY=registry.example.com/my-user make image-local-push deploy` then created some test jobsets, everything worked as expected.

File: README.md [hunk 7]
```diff
@@ -6,8 +6,7 @@ JobSet: An API for managing a group of Jobs as a unit.
 
 To install the CRD and deploy the controller on the cluster selected on your `~/.kubeconfig`:
 
-1. `make docker-build docker-push IMG=<some-registry>/<project-name>:tag`
-2. `make deploy IMG=<some-registry>/<project-name>:tag`
+`IMAGE_REGISTRY=<registry>/<project> image-local-push deploy`
```
Comment (ahg-g) [inline README.md:10]: Should call make?

File: README.md [hunk 8]
```diff
@@ -6,8 +6,7 @@ JobSet: An API for managing a group of Jobs as a unit.
...
```
Comment (ahg-g) [inline README.md:9]:
```suggestion
To build JobSet from source and install JobSet in your cluster, run the following commands:

git clone https://github.com/kubernetes-sigs/jobset.git
cd jobset
IMAGE_REGISTRY=registry.example.com/my-user make image-local-push deploy
```

File: README.md [hunk 9]
```diff
@@ -6,8 +6,7 @@ JobSet: An API for managing a group of Jobs as a unit.
 
 To install the CRD and deploy the controller on the cluster selected on your `~/.kubeconfig`:
```
Comment (ahg-g) [inline README.md:7]:
```suggestion
To build JobSet from source and install JobSet in the cluster selected on your `~/.kubeconfig`, run the following commands:
```

File: Makefile [hunk 10]
```diff
@@ -117,8 +136,8 @@ uninstall: manifests kustomize ## Uninstall CRDs from the K8s cluster specified
 
 .PHONY: deploy
 deploy: manifests kustomize ## Deploy controller to the K8s cluster specified in ~/.kube/config.
-	cd config/manager && $(KUSTOMIZE) edit set image controller=${IMG}
-	$(KUSTOMIZE) build config/default | kubectl apply -f -
+	cd config/manager && $(KUSTOMIZE) edit set image controller=${IMAGE_TAG}
+	$(KUSTOMIZE) build config/default | kubectl create -f -
```
[the remark this answers is not in the document]
Comment (danielvegamyhre) [reply Makefile]: Great find -it works! I'll mark #47 as fixed

**Pull request merged**